tests/test_framework_runtime_adapters.py

```
@@ -303,6 +313,72 @@ def __init__(self):
     assert attention.linear_qkv.allreduce_dgrad is False
 
 
+def test_megatron_te_attention_projection_uses_injected_strict_tp_reduce():
+    calls: list[torch.Tensor] = []
+
+    class ColumnLinear:
+        def __init__(self):
+            self.layer_norm_weight = torch.ones(2)
+            self.weight = torch.tensor([[1.0, 0.0], [0.0, 1.0]])
+
+        def _forward_impl(self, input, weight, *args, **kwargs):
```

Ruff (0.16.3)

[error] Function argument `input` is shadowing a Python builtin

(A002)